File: packages/tools/src/tools/edit.test.ts
import { test } from "node:test";
import { strictEqual, ok } from "node:assert";
import { readFile, writeFile, mkdir, rm } from "node:fs/promises";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { editTool } from "./edit.js";
import type { ToolContext } from "@forge/types";

const testDir = join(tmpdir(), `forge-edit-test-${Date.now()}`);

// Helper to create test context
function makeCtx(): ToolContext {
  return {
    cwd: testDir,
    sessionId: "test-session",
    threadId: "test-thread",
    signal: new AbortController().signal,
    emit: () => {},
  };
}

test("Edit: replaces unique string", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "test.txt");
  await writeFile(file, "Hello Troy Barnes!\nWelcome to Greendale.\n");

  const result = await editTool.handler(
    {
      file_path: file,
      old_string: "Troy Barnes",
      new_string: "Abed Nadir",
    },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);
  ok(result.content[0].type === "text");
  ok(result.content[0].text.includes("replaced"));

  const contents = await readFile(file, "utf-8");
  strictEqual(contents, "Hello Abed Nadir!\nWelcome to Greendale.\n");

  await rm(testDir, { recursive: true });
});

test("Edit: returns error when old_string not found", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "test.txt");
  await writeFile(file, "Hello Troy Barnes!\n");

  const result = await editTool.handler(
    {
      file_path: file,
      old_string: "Jeff Winger",
      new_string: "Dean Pelton",
    },
    makeCtx(),
  );

  strictEqual(result.isError, true);
  ok(result.content[0].type === "text");
  ok(result.content[0].text.includes("not found"));

  await rm(testDir, { recursive: true });
});

test("Edit: returns error when old_string appears multiple times without replace_all", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "test.txt");
  await writeFile(
    file,
    "Troy loves Troy. Troy is Troy. Troy Troy Troy.\n",
  );

  const result = await editTool.handler(
    {
      file_path: file,
      old_string: "Troy",
      new_string: "Abed",
    },
    makeCtx(),
  );

  strictEqual(result.isError, true);
  ok(result.content[0].type === "text");
  ok(result.content[0].text.includes("appears 7 times"));
  ok(result.content[0].text.includes("not unique"));

  await rm(testDir, { recursive: true });
});

test("Edit: replaces all occurrences when replace_all is true", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "test.txt");
  await writeFile(
    file,
    "Troy loves Troy. Troy is Troy. Troy Troy Troy.\n",
  );

  const result = await editTool.handler(
    {
      file_path: file,
      old_string: "Troy",
      new_string: "Abed",
      replace_all: true,
    },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);
  ok(result.content[0].type === "text");
  ok(result.content[0].text.includes("7 occurrences"));

  const contents = await readFile(file, "utf-8");
  strictEqual(
    contents,
    "Abed loves Abed. Abed is Abed. Abed Abed Abed.\n",
  );

  await rm(testDir, { recursive: true });
});

test("Edit: handles multiline replacements", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "test.txt");
  await writeFile(
    file,
    "function hello() {\n  console.log('old');\n}\n",
  );

  const result = await editTool.handler(
    {
      file_path: file,
      old_string: "function hello() {\n  console.log('old');\n}",
      new_string: "function hello() {\n  console.log('new');\n}",
    },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);

  const contents = await readFile(file, "utf-8");
  strictEqual(
    contents,
    "function hello() {\n  console.log('new');\n}\n",
  );

  await rm(testDir, { recursive: true });
});

test("Edit: returns error for nonexistent file", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "nonexistent.txt");

  const result = await editTool.handler(
    {
      file_path: file,
      old_string: "foo",
      new_string: "bar",
    },
    makeCtx(),
  );

  strictEqual(result.isError, true);
  ok(result.content[0].type === "text");
  ok(result.content[0].text.includes("does not exist"));

  await rm(testDir, { recursive: true });
});

test("Edit: preserves file when old_string and new_string are identical", async () => {
  await mkdir(testDir, { recursive: true });
  const file = join(testDir, "test.txt");
  const original = "Human Being mascot\n";
  await writeFile(file, original);

  const result = await editTool.handler(
    {
      file_path: file,
      old_string: "mascot",
      new_string: "mascot",
    },
    makeCtx(),
  );

  strictEqual(result.isError, undefined);

  const contents = await readFile(file, "utf-8");
  strictEqual(contents, original);

  await rm(testDir, { recursive: true });
});

File: internal/tools/edit_test.go
package tools

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/jelmersnoeck/forge/internal/types"
	"github.com/stretchr/testify/require"
)

func TestEditTool(t *testing.T) {
	tests := map[string]struct {
		setup func(t *testing.T, dir string) map[string]any
		want  func(*testing.T, string, types.ToolResult, error)
	}{
		"single replacement": {
			setup: func(t *testing.T, dir string) map[string]any {
				path := filepath.Join(dir, "script.txt")
				err := os.WriteFile(path, []byte("Troy: Cool beans"), 0644)
				require.NoError(t, err)
				return map[string]any{
					"file_path":  path,
					"old_string": "Cool",
					"new_string": "Awesome",
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				r.Contains(result.Content[0].Text, "replaced 1 occurrence")

				path := filepath.Join(dir, "script.txt")
				data, readErr := os.ReadFile(path)
				r.NoError(readErr)
				r.Equal("Troy: Awesome beans", string(data))
			},
		},
		"replace all": {
			setup: func(t *testing.T, dir string) map[string]any {
				path := filepath.Join(dir, "chant.txt")
				err := os.WriteFile(path, []byte("pop pop pop"), 0644)
				require.NoError(t, err)
				return map[string]any{
					"file_path":   path,
					"old_string":  "pop",
					"new_string":  "bang",
					"replace_all": true,
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)
				r.Contains(result.Content[0].Text, "replaced 3 occurrence")

				path := filepath.Join(dir, "chant.txt")
				data, readErr := os.ReadFile(path)
				r.NoError(readErr)
				r.Equal("bang bang bang", string(data))
			},
		},
		"multiple occurrences without replace_all": {
			setup: func(t *testing.T, dir string) map[string]any {
				path := filepath.Join(dir, "dean.txt")
				err := os.WriteFile(path, []byte("Dean Dean Dean"), 0644)
				require.NoError(t, err)
				return map[string]any{
					"file_path":  path,
					"old_string": "Dean",
					"new_string": "Pelton",
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.True(result.IsError)
				r.Contains(result.Content[0].Text, "appears 3 times")
				r.Contains(result.Content[0].Text, "replace_all")
			},
		},
		"old_string not found": {
			setup: func(t *testing.T, dir string) map[string]any {
				path := filepath.Join(dir, "study.txt")
				err := os.WriteFile(path, []byte("Study group meeting"), 0644)
				require.NoError(t, err)
				return map[string]any{
					"file_path":  path,
					"old_string": "Pierce",
					"new_string": "Leonard",
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.True(result.IsError)
				r.Contains(result.Content[0].Text, "not found")
			},
		},
		"file not found": {
			setup: func(t *testing.T, dir string) map[string]any {
				return map[string]any{
					"file_path":  filepath.Join(dir, "missing.txt"),
					"old_string": "foo",
					"new_string": "bar",
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.True(result.IsError)
				r.Contains(result.Content[0].Text, "file not found")
			},
		},
		"multiline replacement": {
			setup: func(t *testing.T, dir string) map[string]any {
				path := filepath.Join(dir, "lyrics.txt")
				content := "Troy and Abed\nin the morning"
				err := os.WriteFile(path, []byte(content), 0644)
				require.NoError(t, err)
				return map[string]any{
					"file_path":  path,
					"old_string": "Troy and Abed\nin the morning",
					"new_string": "Troy and Abed\nin the evening",
				}
			},
			want: func(t *testing.T, dir string, result types.ToolResult, err error) {
				r := require.New(t)
				r.NoError(err)
				r.False(result.IsError)

				path := filepath.Join(dir, "lyrics.txt")
				data, readErr := os.ReadFile(path)
				r.NoError(readErr)
				r.Equal("Troy and Abed\nin the evening", string(data))
			},
		},
	}

	for name, tc := range tests {
		t.Run(name, func(t *testing.T) {
			dir := t.TempDir()
			input := tc.setup(t, dir)
			tool := EditTool()
			result, err := tool.Handler(input, types.ToolContext{})
			tc.want(t, dir, result, err)
		})
	}
}
